hazelcast/src/main/java/com/hazelcast/dataconnection/impl/JdbcDataConnection.java

```diff
@@ -51,27 +52,32 @@ public class JdbcDataConnection extends DataConnectionBase {
 
     private static final AtomicInteger DATA_SOURCE_COUNTER = new AtomicInteger();
 
-    private HikariDataSource pooledDataSource;
+    private volatile ConcurrentMemoizingSupplier<HikariDataSource> pooledDataSourceSup;
     private Supplier<Connection> singleUseConnectionSup;
 
     public JdbcDataConnection(DataConnectionConfig config) {
```

Comment:
IMO we should document explicilty in `DataConnection` javadoc that the constructor is not allowed to throw under any circumstances

Reply:
45e4692c553

@Fly-Style already has a PR to fix this for Kafka.

@ldziedziul is working on a fix for HazelcastDataLink

@TomaszGaweda could you verify that the MongoDataConnection behaves according to the updated Javadoc?

Reply:
@frant-hartm created https://github.com/hazelcast/hazelcast/pull/24355